modules/dists/BlockDist.chpl

```diff
@@ -319,7 +319,215 @@ This example demonstrates a Block-distributed sparse domain and array:
+      // we use "method call resolves" instead of just "resolves".
+      if __primitive("method call resolves", _value, "dsiNewRectangularDom",
+                     rank, idxType, strides, ranges) {
+        return _value.dsiNewRectangularDom(rank, idxType, strides, ranges);
+      }
+
+      // The following supports deprecation by Vass in 1.31 to implement #17131
+      // Once range.stridable is removed, replace chpl_dsiNRDhelp() with
+      //   var x = _value.dsiNewRectangularDom(..., strides, ranges);
+      // and uncomment proc dsiNewRectangularDom() in ChapelDistribution.chpl
+
+      param stridable = strides.toStridable();
+      const ranges2 = chpl_convertRangeTuple(ranges, stridable);
+      if __primitive("method call resolves", _value, "dsiNewRectangularDom",
+                     rank, idxType, stridable, ranges2) {
+
+        compilerWarning("the domain map '", _value.type:string,
+          "' needs to be updated from 'stridable: bool' to",
+          " 'strides: strideKind' because 'stridable' is deprecated");
+
+        return _value.dsiNewRectangularDom(rank, idxType, stridable, ranges2);
+      }
+
+      compilerError("rectangular domains are not supported by",
+                    " the distribution ", this.type:string);
+    }
+
+    proc newRectangularDom(param rank: int, type idxType,
+                           param strides: strideKind,
+                           ranges: rank*range(idxType, boundKind.both, strides),
+                           definedConst: bool = false) {
+      var x = chpl_dsiNRDhelp(rank, idxType, strides, ranges);
+
+      x.definedConst = definedConst;
+
+      if x.linksDistribution() {
+        _value.add_dom(x);
+      }
+      return x;
+    }
+
+    proc newRectangularDom(param rank: int, type idxType,
+                           param strides: strideKind,
+                           definedConst: bool = false) {
+      var ranges: rank*range(idxType, boundKind.both, strides);
+      return newRectangularDom(rank, idxType, strides, ranges, definedConst);
+    }
+
+    proc idxToLocale(ind) do return _value.dsiIndexToLocale(ind);
+
+    proc writeThis(f) throws {
+      f.write(_value);
+    }
+
+    @chpldoc.nodoc
+    proc serialize(writer, ref serializer) throws {
+      writer.write(_value);
+    }
+
+    proc displayRepresentation() { _value.dsiDisplayRepresentation(); }
+
+    /*
+       Return an array of locales over which this distribution was declared.
+    */
+    proc targetLocales() const ref {
+      return _value.dsiTargetLocales();
+    }
+
+  @chpldoc.nodoc
+  inline operator ==(d1: Block(?), d2: Block(?)) {
+    if (d1._value == d2._value) then
+      return true;
+    return d1._value.dsiEqualDMaps(d2._value);
+  }
+
+  @chpldoc.nodoc
+  inline operator !=(d1: Block(?), d2: Block(?)) {
+    return !(d1 == d2);
+  }
+}
+
+
+
+@chpldoc.nodoc
+@unstable(category="experimental", reason="assignment between distributions is currently unstable due to lack of testing")
```

Comment:
Not about this PR, but I wonder whether we really thought about semantics of distribution assignment. Does it redistribute the domain(s) that use this distribution? Doesn't look like it by looking at the implementation here, but I thought we tracked domains of a distribution similar to tracking arrays of a domain..

Reply:
It doesn't, though it ought to.  We did used to track domains of a distribution, but stopped at some point when it became too much of a performance bottleneck since we weren't using it.  This came up most recently in #22129 where @ronawho reminded me of this history and suggested that we could switch to the same doubly-linked list implementation that we're currently using to track arrays within domains and is sufficiently low-overhead.  At the time that we ripped out the previous dist->domain tracking, we had not yet performed that optimization.

Reply:
I think https://github.com/chapel-lang/chapel/issues/22568 captures this